proposals/003-metric-improvenents.md

```diff
@@ -0,0 +1,262 @@
+  hard to use this metric to understand something about produce requests or fetch responses.
+* *1 / *2 - we have contradictory meanings for the `flowing` label.  in *2, requests flow `UPSTREAM`, responses flow `DOWNSTREAM`.
+  In *1 requests flow `DOWNSTREAM` which makes no sense.
+* *3 - downstream errors does not record account for errors such as TLS negotiation errors or failures to determine virtual cluster or broker.  
+
+### General short-comings/weaknesses
+
+* The metrics endpoint populates `HELP` for each metric, but this is populated with nothing more than the metric's name.
+  We ought to be giving a clear description of the metric.
+* Some of the label names don't follow Prometheus naming conventions.  Prometheus recommends snake case whereas we have used camel case in some cases.
+
+## Use Cases
+
+In general, metrics exist to help a user understand the health and performance of the proxy. They help inform
+capacity planning, drive system alerts reporting abnormal conditions and help with problem-solving (things like
+diagnosing failures or system slow-downs).
+
+Introducing a proxy to a Kafka system will increase latency of the system. Users will need a mechanism to understand how
+much latency is being added for their particular use-cases and have the ability to monitor this latency over time.
+
+Kroxylicious Filters themselves have the potential to alter behaviour of the Kafka system.  Users will need insights into
+performance and other issues that might result from filter behaviour. Specifically:
+
+* filters have the ability to change the size of the RPCs as they traverse the proxy (record encryption)
+* filters have the ability to change the flow of messages within the Kafka system. Specifically, they can short-circuit
+  requests by producing their own responses (the broker never sees the request) and introduce their own requests.
+* filters can perform asynchronous work.  When asynchronous work occurs, the proxy stops reading the downstream or upstream.
+
+## Proposal
+
+## Summary
+
+The following metrics will be added.  They will each be described in more detail later in this section.
+
+| Metric                                           | Type         | Labels                                                                                | Description                                                                             |
+|--------------------------------------------------|--------------|---------------------------------------------------------------------------------------|-----------------------------------------------------------------------------------------|
+| kroxylicious_downstream_kafka_message_total      | Counter      | virtual_cluster, node_id†, flowing, api_key, api_version, opaque                      | Incremented by one every time a request comes from/response goes to the downstream.     |
```

Comment:
Why should we have `flowing` as a label? I think people are actually interested in requests and responses, which are fundamentally different things. Having it as a label means to get when you want you need to filter as well as aggregate.

To me this should be two separate metrics:

* `client_to_proxy_request_total`
* `proxy_to_client_response_total`

These names are self-explanatory, and shorter.

Reply:
Yes, i think this is better.  I'll make a change.